### ISSUE TITLE
weekDay	is always 0

### ISSUE THREAD

**Issue by donsenilo1968** [RobTillaart/DS3232#2]
Hi

Even in your Example (with added weekDay) it always returns 0.

**Comment by RobTillaart** [comment]
Thanks for the issue.
Might take some time before I can investigate.

**Comment by RobTillaart** [comment]
What platform do you use?
What is the output of **DS3232_demo_read.ino**?
Do you have pull up resistors?
What is the wire length?

**Comment by donsenilo1968** [comment]
Arduino Nano
Date and Time
No
Max. 3cm

**Comment by RobTillaart** [comment]
So date and time works but not weekdays ....

Need a code / datasheet  check as day of week is iirc the only read only register..

**Comment by donsenilo1968** [comment]
Yep. Weekdays are always 0.
This is the only [Datasheet](https://cdn.shopify.com/s/files/1/1509/1638/files/DS3231_Real_Time_Clock_Datenblatt_AZ-Delivery_Vertriebs_GmbH_e99489a5-e10c-4ae7-8bc4-75ac830fd280.pdf?v=1613398175) I could found.

**Comment by RobTillaart** [comment]
Datasheet from the manufacturer of the device:
- https://www.analog.com/media/en/technical-documentation/data-sheets/DS3231.pdf
- https://www.analog.com/media/en/technical-documentation/data-sheets/DS3232.pdf

**Comment by RobTillaart** [comment]
Registers:
![image](https://github.com/RobTillaart/DS3232/assets/462844/bbf1ba28-94e3-44ea-be8d-6a18838ffc53)

Register 3 is Day of week - 1..7
```cpp
uint8_t DS3232::weekDay() { return _reg[3]; }
```
the register 3 is filled in **read()** and not masked or so. 
The bcd2dec() function that converts the BCD values from the registers is not needed for weekday as it does not change the value.

Can you add a print statement into the read function in ds3232.cpp (line 61) ?
Please post what is read from the registers.

```cpp
int DS3232::read()
{
  _wire->beginTransmission(_address);
  _wire->write(DS3232_SECONDS);
  _rv = _wire->endTransmission();
  if (_rv != 0) return DS3232_ERROR_I2C;

  if (Wire.requestFrom(_address, (uint8_t)7) != 7)
  {
    return DS3232_ERROR_I2C;
  }
  for (int i = 0; i < 7; i++)
  {
    uint8_t tmp = _wire->read();  //  <<  changed
    Serial.println(tmp, HEX);  //  <<  changed
    _reg[i] = bcd2dec(tmp);  //  <<  changed
  }
  _reg[0] &= 0x7F;
  _reg[2] &= 0x3F;
  _lastRead = millis();

  return DS3232_OK;
}
```

**Comment by donsenilo1968** [comment]
I get this:

16031		24-4-5 11:40:2
1
4
40
11
1
5
4
24
18033		24-4-5 11:40:4
1
6
40
11
1
5
4
24

**Comment by RobTillaart** [comment]
Adding the registers / meaning  
```
| register | meaning      | value |
|:--------:|:-------------|:-----:|
|    0     | seconds      |   4   |
|    1     | minutes      |  40   |
|    2     | hours        |  11   |
|    3     | dayofweek    |   1   |
|    4     | day of month |   5   |
|    5     | month        |   4   |
|    6     | year         |  24   |
18033 24-4-5 11:40:4
```

Datasheet states 
_The day-of-week register increments at midnight.
Values that correspond to the day of week are user defined but must be sequential (i.e., if 1 equals
Sunday, then 2 equals Monday, and so on). Illogical
time and date entries result in undefined operation._

The day of week is 1 where we would expect 5 or 6 (?)
As the value is now 1, I assume the RTC started yesterday with a value 0 and it was updated last midnight.

That would mean I need to check the library if the DOW is set.

**Comment by RobTillaart** [comment]
> That would mean I need to check the library if the DOW is set.

Has no setWeekDay() so ==> BUG

**Comment by RobTillaart** [comment]
@donsenilo1968 
Created a develop branch that adds **setWeekDay()** (and some minor edits)
If you have time, can you verify if it works?

update - valid values seem to be 1..7

- https://github.com/RobTillaart/DS3232/tree/develop

**Comment by donsenilo1968** [comment]
With the Develop Branch

Before demo_write:
212216		24-4-5 20:12:25
1
214219		24-4-5 20:12:27
1

After demo_write:
144150		24-4-5 20:15:50
5
146152		24-4-5 20:15:53
5

If I understand it correctly, once the Weekday is set it increments at midnight and starts again on Mondays with 1.

But (in my Case) I use Udp NTP Client to get the actual Date and Time and it seems there is no Way to get the Weekday.

**Comment by RobTillaart** [comment]
That is what I understand too. 
Please let me know the weekday tomorrow to be sure. Then I can merge the new code.

Thanks for testing!

**Comment by donsenilo1968** [comment]
I found [here](https://forum.arduino.cc/t/got-time-from-ntp-server-need-day-of-week/117799/6) an interesting Part about getting the Weekday. If this is true, the Output for the Weekday was correct (1 = Friday).

**Comment by RobTillaart** [comment]
> I found [here](https://forum.arduino.cc/t/got-time-from-ntp-server-need-day-of-week/117799/6) an interesting Part about getting the Weekday. If this is true, the Output for the Weekday was correct (1 = Friday).

Then it could automatically update it to 2 tomorrow.

Indeed an interesting find

**Comment by RobTillaart** [comment]
@donsenilo1968 
Added a note about the Thursday == 0 weekday in the readme.md

**Comment by donsenilo1968** [comment]
Hi Rob

It seems like it works.

![20240407_011236_001](https://github.com/RobTillaart/DS3232/assets/21273937/0a3b8b3d-edf1-49ff-b9e0-4a2583044a6d)

I read the Date and Time via NTP, add the Timezone and calculate the Weekday from the epoch. Then I write all to the Clock (Date and Time at the Bottom of the Display). This happens when the Arduino starts and every Monday.
The Date, Time and Weekday at the Top are the Data getting from the Clock.

Thank you very much for your Help and updating the Code.
Best Regards from Vienna,
Markus

**Comment by RobTillaart** [comment]
Thanks for the photo, looks good!

The NTP time is the moment of the last sync I assume?
Or is there another reason for the 5 minutes difference?

Are there other things missing in the library?
(I am thinking about adding EEPROM support)

**Comment by donsenilo1968** [comment]
_The NTP time is the moment of the last sync I assume?_
Yes, exactly.

I use this Lib because it's minimalistic and I don't need any other Functions for my Project.
Is adding more Functions necessary ? It's your Baby, so it's your Decision.

**Comment by RobTillaart** [comment]
My plan is to create a derived extended class
DS3232-EXT
Which has the additional functions.
Exactly because that keeps the base class minimalistic.

Another option I consider is to have a 2nd class DS3232EE that only handles the EEPROM part of the device. Not derived, complete disjunct as there is no reason to have it merged in one object afaik.
(except convenience)
It could be in same library though

Other functionality like alarms etc would be logically more related to the time keeping.
These should therefor be in the EXT class.

**Comment by RobTillaart** [comment]
Merged develop into master and released 0.4.0 version,
